---
vignette: >
  % \VignetteIndexEntry{Data retrieval}
  % \VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
output:
  html_document:
    mathjax: null   
---

# Setup
```{r, message = FALSE}
library(BioPlex)
library(AnnotationHub)
library(ExperimentHub)
library(ExpressionAtlas)
library(graph)
```

# Data resources

## General

Connect to 
[AnnotationHub](http://bioconductor.org/packages/AnnotationHub):

```{r ahub, message = FALSE}
ah <- AnnotationHub::AnnotationHub()
```

Connect to 
[ExperimentHub](http://bioconductor.org/packages/ExperimentHub):

```{r ehub, message = FALSE}
eh <- ExperimentHub::ExperimentHub()
```

OrgDb package for human:
```{r orgdb, message = FALSE}
ahdb <- AnnotationHub::query(ah, c("orgDb", "Homo sapiens"))
orgdb <- ahdb[[length(ahdb)]]
orgdb
keytypes(orgdb)
```

EnsDb package for human:
```{r ensdb, message = FALSE}
ahdb <- AnnotationHub::query(ah, c("ensDb", "Homo sapiens"))
ensdb <- ahdb[[length(ahdb)]]
ensdb
keytypes(ensdb)
```

## BioPlex PPIs

[Available networks](https://bioplex.hms.harvard.edu/interactions.php) include: 

- BioPlex 293T cells (versions 1.0, 2.0, and 3.0)
- BioPlex HCT116 cells (version 1.0)

Let's get the latest version of the 293T PPI network:

```{r bioplex293T}
bp.293t <- getBioPlex(cell.line = "293T", version = "3.0")
head(bp.293t)
```

```{r bioplexHCT116}
bp.hct116 <- getBioPlex(cell.line = "HCT116", version = "1.0")
head(bp.hct116)
```

### ID mapping

The protein-to-gene mappings from BioPlex (i.e. UNIPROT-to-SYMBOL and UNIPROT-to-ENTREZID) are based on the mappings available from Uniprot

We can update those based on Bioc annotation functionality:  

```{r bioplex-remap}
bp.293t.remapped <- getBioPlex(cell.line = "293T",
                               version = "3.0",
                               remap.uniprot.ids = TRUE)
```

Note: this is currently based on the gene-level mappings in the
orgdb package for human, and those might not be optimal for mapping UniPprot
isoform identifiers (eg. "Q8N7W2-1" and "Q8N7W2-2") used in BioPlex. 
The remapping currently simply drops the dash-N extension and we thus might
want to rather base that at one point on the official 
[UniProt ID mapping file](https://ftp.uniprot.org/pub/databases/uniprot/current_release/knowledgebase/idmapping).

### Data structures for BioPlex PPIs

We can also represent a given version of the BioPlex PPI network for a given
cell line as one big graph where bait and prey relationship are represented 
by directed edges from bait to prey.

```{r bpgraph}
bp.gr <- bioplex2graph(bp.293t)
bp.gr
head(graph::nodeData(bp.gr))
head(graph::edgeData(bp.gr))
```

When starting off with ordinary dfs we'll be losing the ability
to annotate graph-level annotation such as cell line, version, PMID, ...
We might need to work with `DataFrame`s where we have `mcols` and `metadata`.


### PFAM domains

We can easily add PFAM domain annotations to the node metadata:

```{r pfam}
bp.gr <- annotatePFAM(bp.gr, orgdb)
head(graph::nodeData(bp.gr, graph::nodes(bp.gr), "PFAM"))
```

## CORUM complexes

Complete set of
[CORUM](http://mips.helmholtz-muenchen.de/corum/#download) complexes:

```{r corumALL}
all <- getCorum(set = "all", organism = "Human")
dim(all)
colnames(all)
all[1:5, 1:5]
```

Core set of complexes:
```{r corumCore}
core <- getCorum(set = "core", organism = "Human")
dim(core)
```

Complexes with splice variants:
```{r corumSplice}
splice <- getCorum(set = "splice", organism = "Human")
dim(splice)
```

### ID mapping

The protein-to-gene mappings from CORUM (i.e. UNIPROT-to-SYMBOL and UNIPROT-to-ENTREZID) might not be fully up-to-date. 

We can update those based on Bioc annotation functionality:  

```{r corum-remap}
core.remapped <- getCorum(set = "core", 
                          organism = "Human",
                          remap.uniprot.ids = TRUE)
```

### Data structures for CORUM complexes 

We can represent the CORUM complexes as a list of character vectors.
The names of the list are the complex IDs/names, and each element of the list is
a vector of UniProt IDs for each complex.

```{r corum2list}
core.list <- corum2list(core, subunit.id.type = "UNIPROT")
head(core.list)
length(core.list)
```

If we'd like to add metadata in a more structured form to each complex,
a suitable option is a 
[GeneSetCollection](https://bioconductor.org/packages/GSEABase).

We can also represent the CORUM complexes as a list of graph instances,
where all nodes of a complex are connected to all other nodes of that complex
with undirected edges.

```{r corum2glist}
core.glist <- corum2graphlist(core, subunit.id.type = "UNIPROT")
head(core.glist)
length(core.glist)
core.glist[[1]]@graphData
graph::nodeData(core.glist[[1]])
```

Note that we can easily convert a 
[graph](https://bioconductor.org/packages/graph) object into an 
[igraph](https://cran.r-project.org/package=igraph) object using
`igraph::graph_from_graphnel`.

## Transcriptome data

### HEK293 cells

#### GSE122425

[GSE122425](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE122425)

```{r gse122425}
se <- getGSE122425()
se
head(assay(se, "raw"))
head(assay(se, "rpkm"))
colData(se)
rowData(se)
```

### HCT116 cells

#### Cancer Cell Line Encyclopedia (CCLE)

Get RNA-seq data for 934 cancer cell lines (incl. HCT116) from the 
[Cancer Cell Line Encyclopedia](https://portals.broadinstitute.org/ccle) 
as available from the [ArrayExpress-ExpressionAtlas](https://www.ebi.ac.uk/gxa) 
(Accession: [E-MTAB-2770](https://www.ebi.ac.uk/gxa/experiments/E-MTAB-2770))

```{r CCLE}
atlasRes <- ExpressionAtlas::searchAtlasExperiments(
              properties = "Cancer Cell Line Encyclopedia", 
              species = "human" )
atlasRes
ccle <- ExpressionAtlas::getAtlasData(atlasRes$Accession)
ccle <- ccle$`E-MTAB-2770`$rnaseq
ccle
```

We have raw read counts: 
```{r ccle-assay}
assay(ccle)[1:5,1:5]
colData(ccle)
```

and we have one sample per cell line:

```{r ccle-hct116}
ind <- grep("HCT 116", ccle$cell_line)
colData(ccle)[ind,]
```

#### Klijn et al., 2015

Get RNA-seq data of 675 commonly used human cancer cell lines (incl. HCT116) from [Klijn et al., 2015](https://pubmed.ncbi.nlm.nih.gov/25485619) 
as available from the [ArrayExpress-ExpressionAtlas](https://www.ebi.ac.uk/gxa) 
(Accession: [E-MTAB-2706](https://www.ebi.ac.uk/gxa/experiments/E-MTAB-2706))

```{r klijn}
klijn <- ExpressionAtlas::getAtlasData("E-MTAB-2706")
klijn <- klijn$`E-MTAB-2706`$rnaseq
klijn
```

We have raw read counts: 
```{r klijn-assay}
assay(klijn)[1:5,1:5]
colData(klijn)
```

and we have one sample per cell line:

```{r klijn-hct116}
ind <- grep("HCT 116", klijn$cell_line)
colData(klijn)[ind,]
```

Combine HCT-116 transcriptome from CCLE and Klijn et al. datasets:

```{r comb-hct116}
ind1 <- grep("HCT 116", ccle$cell_line)
ind2 <- grep("HCT 116", klijn$cell_line)
emat <- cbind(assay(ccle)[,ind1], assay(klijn)[,ind2])
colnames(emat) <- c("ccle", "klijn")
head(emat)
```

## Proteome data

### CCLE

Pull the data from ExperimentHub:
```{r ccle-proteom, eval = FALSE}
ehdb <- AnnotationHub::query(eh, c("gygi", "depmap"))
ccle.prot <- ehdb[[length(ehdb)]]
ccle.prot <- as.data.frame(ccle.prot)
```

Explore the data:
```{r ccle-proteom2, eval = FALSE}
dim(ccle.prot)
colnames(ccle.prot)
head(ccle.prot)
```

Restrict to HCT116:

```{r ccle-prot-hct116, eval = FALSE}
ccle.prot.hct116 <- subset(ccle.prot, cell_line == "HCT116_LARGE_INTESTINE")
dim(ccle.prot.hct116)
head(ccle.prot.hct116)
```

Or turn into a `SummarizedExperiment` for convenience (we can restrict
this to selected cell lines, but here we keep all cell lines):

```{r ccle-prot-se, eval = FALSE}
se <- ccleProteome2SummarizedExperiment(ccle.prot, cell.line = NULL)
assay(se)[1:5, 1:5]
assay(se)[1:5, "HCT116"]
rowData(se)
```

### Relative protein expression data (preprint)

Huttlin_BioPlex3_Table_S4.xlsx contains the relative protein expression data comparing 293T and HCT116 cells. This is a supplementary table that will be published with the upcoming BioPlex 3 paper. 

That lives on dropbox for the moment, and can be wrapped into an R data
package or uploaded to ExperimentHub. Once published there will be other
avenues to pull the data (web request, reference DB, ...).

```{r bp.prot, eval = FALSE}
bp.prot <- getBioplexProteome()
assay(bp.prot)[1:5,1:5]
colData(bp.prot)
rowData(bp.prot)
```

There are 10 channels - 5 replicates for 293T and 5 for HCT116.
Based on the way we collect these data, we end up viewing them so that they're scaled to add up to 100%.  Essentially, we think of them as ten variably-sized slices of a pie, with five each allotted to each cell line. Modeling the data this way presents some statistical complications, as I'm sure you can imagine, but it most closely aligns with how our data work.

# Caching

Note that calling functions like `getCorum` or `getBioPlex` with argument
`cache = FALSE` will automatically overwrite the corresponding object in your 
cache. It is thus typically not required for a user to interact with the cache.

For more extended control of the cache, use from within R:

```{r cache}
cache.dir <- tools::R_user_dir("BioPlex", which = "cache") 
bfc <- BiocFileCache::BiocFileCache(cache.dir)
```

and then proceed as described in the
[BiocFileCache vignette, Section 1.10](https://www.bioconductor.org/packages/release/bioc/vignettes/BiocFileCache/inst/doc/BiocFileCache.html#cleaning-or-removing-cache)
either via `cleanbfc()` to clean or `removebfc()` to remove your cache.

To do a hard reset (use with caution!):

```{r rmCache, eval = FALSE}
BiocFileCache::removebfc(bfc)
```

## SessionInfo

```{r sessionInfo}
sessionInfo()
```
