---
vignette: >
  % \VignetteIndexEntry{Data retrieval}
  % \VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
output:
  html_document:
    mathjax: null   
---

# Setup
```{r, message = FALSE}
library(BioPlex)
library(AnnotationDbi)
library(ExpressionAtlas)
library(org.Hs.eg.db)
```

# Data resources

## BioPlex PPIs

[Available networks](https://bioplex.hms.harvard.edu/interactions.php) include: 

* BioPlex 293T cells (versions 1.0, 2.0, and 3.0)
* BioPlex HCT116 cells (version 1.0)

Let's get the latest version of the 293T PPI network:

```{r bioplex293T}
bp.293t <- getBioPlex(cell.line = "293T", version = "3.0")
bp.293t
```

```{r bioplexHCT116}
bp.hct116 <- getBioPlex(cell.line = "HCT116", version = "1.0")
bp.hct116
```

## CORUM complexes

Complete set of
[CORUM](http://mips.helmholtz-muenchen.de/corum/#download) complexes:

```{r corumALL}
all <- getCorum(set = "all")
all
```

Core set of complexes:
```{r corumCore}
core <- getCorum(set = "core")
core
```

Complexes with splice variants:
```{r corumSplice}
splice <- getCorum(set = "splice")
splice
```

We can represent the CORUM complexes as list of character vectors.
The names of the list are the complex IDs/names, and each element of the list is a vector of UniProt IDs for each complex.

```{r corum2list}
core.list <- corum2list(core, organism = "Human", subunit.id.type = "UNIPROT")
head(core.list)
length(core.list)
```

If we'd like to add metadata in a more structured form to each complex, a suitable option is a 
[GeneSetCollection](https://bioconductor.org/packages/GSEABase).

## Transcriptome data

### HEK293 cells

#### GSE122425

Get RNA-seq data for HEK293 cells from GEO: 
[GSE122425](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE122425)

```{r gse122425}
se <- getGSE122425()
se
head(assay(se, "raw"))
head(assay(se, "rpkm"))
colData(se)
rowData(se)
```

Sanity check that the expressed genes are showing up as prey:

```{r prey-expression}
bait <- unique(bp.293t$SymbolA)
length(bait)
prey <- unique(bp.293t$SymbolB)
length(prey)
```

```{r}
ind <- match(prey, rowData(se)[["GeneSymbol"]])
par(las = 2)
boxplot(log2(assay(se, "rpkm") + 0.5)[ind,], 
        names = se$title, 
        ylab = "log2 RPKM")
```

How many prey genes are expressed (raw read count > 0) in all 3 WT reps:

```{r prey-expression2}
# background: how many genes in total are expressed in all three WT reps
gr0 <- rowSums(assay(se)[,1:3] > 0)
table(gr0 == 3)
# prey: expressed in all three WT reps
table(gr0[ind] == 3)
# prey: expressed in at least one WT rep
table(gr0[ind] > 0)
```

Are prey genes overrepresented in the expressed genes?
```{r prey-expression-ora}
exprTable <-
     matrix(c(9346, 1076, 14717, 32766),
            nrow = 2,
            dimnames = list(c("Expressed", "Not.expressed"),
                            c("In.prey.set", "Not.in.prey.set")))
exprTable
```

Test using hypergeometric test (i.e. one-sided Fisher's exact test):
```{r prey-expression-ora2}
fisher.test(exprTable, alternative = "greater")
```

Alternatively: permutation test, i.e. repeatedly sample number of prey genes 
from the background, and assess how often we have as many or more than 9346 genes
expressed:

```{r prey-expression-293T-perm}
permgr0 <- function(gr0, nr.genes = length(prey)) 
{
    ind <- sample(seq_along(gr0), nr.genes)
    sum(gr0[ind] == 3)
}
```

```{r prey-expression-perm2}
perms <- replicate(permgr0(gr0), 1000)
summary(perms)
(sum(perms >= 9346) + 1) / 1001
```

### HCT116 cells

#### Cancer Cell Line Encyclopedia (CCLE)

Get RNA-seq data for 934 cancer cell lines (incl. HCT116) from the 
[Cancer Cell Line Encyclopedia](https://portals.broadinstitute.org/ccle) 
as available from the [ArrayExpress-ExpressionAtlas](https://www.ebi.ac.uk/gxa) 
(Accession: [E-MTAB-2770](https://www.ebi.ac.uk/gxa/experiments/E-MTAB-2770))

```{r CCLE}
atlasRes <- ExpressionAtlas::searchAtlasExperiments(
              properties = "Cancer Cell Line Encyclopedia", 
              species = "human" )
atlasRes
ccle <- ExpressionAtlas::getAtlasData(atlasRes$Accession)
ccle <- ccle$`E-MTAB-2770`$rnaseq
ccle
```

We have raw read counts: 
```{r ccle-assay}
assay(ccle)[1:5,1:5]
colData(ccle)
```

and we have one sample per cell line:

```{r ccle-hct116}
ind <- grep("HCT 116", ccle$cell_line)
colData(ccle)[ind,]
```

#### Klijn et al., 2015

Get RNA-seq data of 675 commonly used human cancer cell lines (incl. HCT116) from [Klijn et al., 2015](https://pubmed.ncbi.nlm.nih.gov/25485619) 
as available from the [ArrayExpress-ExpressionAtlas](https://www.ebi.ac.uk/gxa) 
(Accession: [E-MTAB-2706](https://www.ebi.ac.uk/gxa/experiments/E-MTAB-2706))

```{r klijn}
klijn <- ExpressionAtlas::getAtlasData("E-MTAB-2706")
klijn <- klijn$`E-MTAB-2706`$rnaseq
klijn
```

We have raw read counts: 
```{r klijn-assay}
assay(klijn)[1:5,1:5]
colData(klijn)
```

and we have one sample per cell line:

```{r klijn-hct116}
ind <- grep("HCT 116", klijn$cell_line)
colData(klijn)[ind,]
```

Sanity check that the expressed genes are showing up as prey:

```{r prey-expression-hct116}
bait <- unique(bp.hct116$GeneA)
length(bait)
prey <- unique(bp.hct116$GeneB)
length(prey)
```

Map prey to ENSEMBL gene IDs:

```{r mapIDs-hct116}
prey <- AnnotationDbi::mapIds(org.Hs.eg.db, 
                              keytype = "ENTREZID", 
                              column = "ENSEMBL",
                              keys = as.character(prey))
prey <- unlist(unname(prey))
ind.prey <- match(prey, rownames(ccle))
```

Combine HCT-116 transcriptome from CCLE and Klijn et al. datasets:

```{r comb-hct116}
ind1 <- grep("HCT 116", ccle$cell_line)
ind2 <- grep("HCT 116", klijn$cell_line)
emat <- cbind(assay(ccle)[,ind1], assay(klijn)[,ind2])
colnames(emat) <- c("ccle", "klijn")
head(emat)
```

The CCLE and Klijn et al. measurements correlate very well:

```{r cor-ccle-klijn}
cor(emat[,"ccle"], emat[,"klijn"])
```

How many prey genes are expressed (raw read count > 0) in both CCLE and Klijn et al.?

```{r prey-expression2-hct116}
# background: how many genes in total are expressed in both measurements
gr0 <- rowSums(emat > 0)
table(gr0 == 2)
# prey: expressed in both measurements
table(gr0[ind.prey] == 2)
# prey: expressed in at least one measurement
table(gr0[ind.prey] > 0)
```

# Downstream analysis

## BioNet

## GGEA

## SGSeq
